### PR TITLE
Rewrite as a standalone script

### PR DESCRIPTION
**Pull request by weierophinney**
This patch is intended for a v3 release, and rewrites the component to be a standalone script, with no dependencies.

This means:

- Removal of existing classes, which depended on zend-mvc &lt; - &gt; zend-console integration.
- Creation of individual classes for each command, along with *unit tests* (for the first time!).
- Addition of a "status" command for determining if the project has development mode enabled or not.

Finally, because neither ZendDevelopmentTools nor zftool are ready for the zend-mvc v3 release, the README no longer recommends them, nor do the default configuration distributions include them.

Usage now becomes:

```bash
$ ./vendor/bin/zf-development-mode -h      # display a help message
$ ./vendor/bin/zf-development-mode --help  # display a help message
$ ./vendor/bin/zf-development-mode enable  # enable development mode
$ ./vendor/bin/zf-development-mode disable # disable development mode
$ ./vendor/bin/zf-development-mode status  # indicate if enabled or not
```

This can then be included by default in the re-written skeleton application.

### CODE REVIEW

File: src/ConfigDiscoveryTrait.php [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    /**
+     * Return the configured configuration cache directory, if any.
+     *
+     * @return null|string
+     */
+    public function getConfigCacheDir()
+    {
+        if ($this->configCacheDir) {
+            return $this->configCacheDir;
+        }
```
Comment (webimpress) [inline src/ConfigDiscoveryTrait.php]: what is the point of cached config dir? the method is used only once

File: src/ConfigDiscoveryTrait.php [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    /**
+     * Return the configured configuration cache key, if any.
+     *
+     * @return null|string
+     */
+    private function getConfigCacheKey()
+    {
+        if ($this->configCacheKey) {
+            return $this->configCacheKey;
+        }
```
Comment (webimpress) [inline src/ConfigDiscoveryTrait.php]: the same as for `configCacheDir`, useless, method used once

File: src/ConfigDiscoveryTrait.php [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+     * @throws RuntimeException if config/application.config.php does not
+     *     return an array
+     */
+    function getApplicationConfig()
+    {
+        if ($this->applicationConfig) {
+            return $this->applicationConfig;
+        }
+
+        $applicationConfig = (isset($this->projectDir) && ! empty($this->projectDir))
```
Comment (webimpress) [inline src/ConfigDiscoveryTrait.php]: redundant `isset` and `!empty`

File: src/ConfigDiscoveryTrait.php [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+     *
+     * Raises an exception if retrieved configuration is not an array.
+     *
+     * @return array
+     * @throws RuntimeException if config/application.config.php does not
+     *     return an array
+     */
+    function getApplicationConfig()
+    {
+        if ($this->applicationConfig) {
```
Comment (webimpress) [inline src/ConfigDiscoveryTrait.php]: it's not enough (when in previous execution set `$this-> applicationConfig = []` then this condition will be not satisfied - so checks below will be executed for each call):
https://3v4l.org/M0K41

File: src/ConfigDiscoveryTrait.php [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+        $applicationConfig = (isset($this->projectDir) && ! empty($this->projectDir))
+            ? sprintf('%s/%s', $this->projectDir, $this->applicationConfigPath)
+            : $this->applicationConfigPath;
+        if (! file_exists($applicationConfig)) {
+            $this->applicationConfig = [];
+            return $this->applicationConfig;
+        }
+
+        $config = include $applicationConfig;
```
Comment (webimpress) [inline src/ConfigDiscoveryTrait.php]: I think it should be:
`$this->applicationConfig = include $applicationConfig;`

variable and property name are confusing - one is file one is array...

File: src/Disable.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    private $errorStream;
+
+    /**
+     * @param string Path to project.
+     */
+    private $projectDir;
+
+    /**
+     * @param string $projectDir Location to resolve project from.
```
Comment (webimpress) [inline src/Disable.php]: missing `$errorStream` param in doc-block

File: src/Enable.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    private $errorStream;
+
+    /**
+     * @param string Path to project.
+     */
+    private $projectDir;
+
+    /**
+     * @param string $projectDir Location to resolve project from.
```
Comment (webimpress) [inline src/Enable.php]: missing `$errorStream` param in doc-block

File: bin/zf-development-mode [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * @link      http://github.com/zfcampus/zf-development-mode for the canonical source repository
+ * @copyright Copyright (c) 2014-2016 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   http://framework.zend.com/license/new-bsd New BSD License
+ */
+
+namespace ZF\DevelopmentMode;
+
+// Setup/verify autoloading
+if (file_exists($a = dirname(__FILE__).'/../../../autoload.php') === true) {
```
Comment (webimpress) [inline bin/zf-development-mode]: a single space around concatenation operator (is it not specified in PSR-2? I can see we have it in other places)

why we need `=== true` here ??

Comment (webimpress) [inline bin/zf-development-mode]: isn't `dirname(__FILE__)` the same as `__DIR__` ?

File: src/Disable.php (hunk 6 above)
Comment (weierophinney) [inline src/Disable.php]: Added, thanks!

File: src/Enable.php (hunk 7 above)
Comment (weierophinney) [inline src/Enable.php]: Added, thanks!

File: src/Disable.php [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+     * @param null|resource Stream to which to write errors; defaults to STDERR
```
Comment (webimpress) [inline src/Disable.php]: sorry, there is missing the param `$errorStream` name in the doc-block...

File: src/ConfigDiscoveryTrait.php [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+    /**
+     * Return the configured configuration cache directory, if any.
+     *
+     * @return null|string
+     */
+    public function getConfigCacheDir()
+    {
+        $config = $this->getApplicationConfig();
+        if (! isset($config['module_listener_options']['cache_dir'])) {
```
Comment (webimpress) [inline src/ConfigDiscoveryTrait.php]: I think it's better to use `empty`, have you seen my reply:
https://github.com/zendframework/ZendSkeletonApplication/pull/345/files/0fabbf70f7f5559c59af048b57a091a31ca0cfde#r68111033 ?

Comment (weierophinney) [inline src/ConfigDiscoveryTrait.php]: `empty()` will raise an error if any one of the keys in the hierarchy is missing, while `isset()` does not. `isset()` will also return `false` if the value in that key is empty, which makes it more useful for this case.

File: src/Disable.php (hunk 9 above)
Comment (weierophinney) [inline src/Disable.php]: I'll fix on master; I just pushed the tag.

File: src/ConfigDiscoveryTrait.php (hunk 10 above)
Comment (webimpress) [inline src/ConfigDiscoveryTrait.php]: @weierophinney I'm sorry, but you are wrong.
`empty()` never raise any error even if any one of the keys in the hierarchy is missing.
Please check this again, verify on your environment, I'm 100% sure.
Here you have one more example:
https://3v4l.org/2p9RJ

And this is from the PHP documentation:
http://php.net/empty
> No warning is generated if the variable does not exist. That means `empty()` is essentially the concise equivalent to `!isset($var) || $var == false`.

File: src/Status.php [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    /**
+     * Indicate whether or not development mode is enabled.
+     *
+     * @return int
+     */
+    public function __invoke()
+    {
+        if (file_exists($this->develConfigFile)) {
+            // nothing to do
```
Comment (akrabat) [inline src/Status.php]: Comment makes no sense.

**Comment by akrabat** [comment]
Was afk all week. +1 from me.

**Pull request closed without merging**